pkg/runtime/runtime_test.go

```diff
@@ -1290,40 +1290,84 @@ func TestMetadataAppID(t *testing.T) {
 		},
 	}
 
-	pubsubComponent.Spec.Metadata = append(
-		pubsubComponent.Spec.Metadata,
-		commonapi.NameValuePair{
-			Name: "clientID",
-			Value: commonapi.DynamicValue{
-				JSON: v1.JSON{
-					Raw: []byte("{appID} {appID}"),
+	// ClientID should be namespace for k8s
+	t.Run("Kubernetes Mode AppID", func(t *testing.T) {
+		t.Setenv("NAMESPACE", "test")
+		pubsubComponent.Spec.Metadata = append(
+			pubsubComponent.Spec.Metadata,
+			commonapi.NameValuePair{
+				Name: "clientID",
+				Value: commonapi.DynamicValue{
+					JSON: v1.JSON{
+						Raw: []byte("{namespace}"),
+					},
 				},
+			})
+
+		rt, err := NewTestDaprRuntimeWithID(modes.KubernetesMode, "myApp")
+		require.NoError(t, err)
+
+		rt.runtimeConfig.id = daprt.TestRuntimeConfigID
+		defer stopRuntime(t, rt)
+		mockPubSub := new(daprt.MockPubSub)
+
+		rt.runtimeConfig.registry.PubSubs().RegisterComponent(
+			func(_ logger.Logger) pubsub.PubSub {
+				return mockPubSub
 			},
-		})
-	rt, _ := NewTestDaprRuntime(modes.KubernetesMode)
-	rt.runtimeConfig.id = daprt.TestRuntimeConfigID
-	defer stopRuntime(t, rt)
-	mockPubSub := new(daprt.MockPubSub)
+			"mockPubSub",
+		)
 
-	rt.runtimeConfig.registry.PubSubs().RegisterComponent(
-		func(_ logger.Logger) pubsub.PubSub {
-			return mockPubSub
-		},
-		"mockPubSub",
-	)
+		mockPubSub.On("Init", mock.Anything).Return(nil).Run(func(args mock.Arguments) {
+			metadata := args.Get(0).(pubsub.Metadata)
+			clientID := metadata.Properties["clientID"]
+			assert.Equal(t, "test.myApp", clientID)
```

Comment:
Because this method is invoked on a background goroutine, we shouldn't do assertions in here. Assertions in Go should only be on the goroutine where the test is running. Additionally, the test function may return before the callback is executed.

Instead, you should:

1. Create a channel like `clientIDChan := make(chan string)` outside of this `mockPubSub.On`
2. In the callback, write the client ID to the channel: `clientIDChan <- clientID`
3. After `assert.NoError` on L1328, read from the channel and ensure the value matches, such as `assert.Equal(t, "test.myApp", <-clientIDChan)`